src/gui/modals/FileDialog.cpp

```diff
@@ -23,127 +23,79 @@
-					   const QString &directory, const QString &filter ) :
-	QFileDialog( parent, caption, directory, filter )
+FileDialog::FileDialog(QWidget *parent, const QString &caption,
+					   const DirType dirType, 
+					   const QString &filter,
+					   const QString &directory):
+	QFileDialog(parent, caption, directory.isEmpty() ? getPath(dirType) : directory, filter),
+	m_dirType(dirType)
 {
 #if QT_VERSION > 0x050200
 	setOption( QFileDialog::DontUseCustomDirectoryIcons );
 #endif
-
 	setOption( QFileDialog::DontUseNativeDialog );
+}
 
-#ifdef LMMS_BUILD_LINUX
-	QList<QUrl> urls;
-#else
-	QList<QUrl> urls = sidebarUrls();
-#endif
-
-	QDir desktopDir;
-	desktopDir.setPath(QStandardPaths::writableLocation(QStandardPaths::DesktopLocation));
-	if (desktopDir.exists())
-	{
-		urls << QUrl::fromLocalFile(desktopDir.absolutePath());
-	}
-	
-	QDir downloadDir(QDir::homePath() + "/Downloads");
-	if (!downloadDir.exists())
-	{
-		downloadDir.setPath(QStandardPaths::writableLocation(QStandardPaths::DownloadLocation));
-	}
-	if (downloadDir.exists())
-	{
-		urls << QUrl::fromLocalFile(downloadDir.absolutePath());
-	}
-
-	QDir musicDir;
-	musicDir.setPath(QStandardPaths::writableLocation(QStandardPaths::MusicLocation));
-	if (musicDir.exists())
+FileDialog::~FileDialog()
+{
+	if (result() == QDialog::Accepted)
 	{
-		urls << QUrl::fromLocalFile(musicDir.absolutePath());
+		s_lastUsedPaths[m_dirType] = directory().absolutePath();
 	}
+}
 
-	urls << QUrl::fromLocalFile(ConfigManager::inst()->workingDir());
-	
-	// Add `/Volumes` directory on OS X systems, this allows the user to browse
-	// external disk drives.
-#ifdef LMMS_BUILD_APPLE
-	QDir volumesDir( QDir("/Volumes") );
-	if ( volumesDir.exists() )
-		urls << QUrl::fromLocalFile( volumesDir.absolutePath() );
-#endif
-
-#ifdef LMMS_BUILD_LINUX
-
-	// FileSystem types : https://www.javatpoint.com/linux-file-system
-	QStringList usableFileSystems = {"ext", "ext2", "ext3", "ext4", "jfs", "reiserfs", "ntfs3", "fuse.sshfs", "fuseblk"};
+QString FileDialog::getDefaultPath(const DirType dirType)
+{
+	auto* config = ConfigManager::inst();
 
-	for(QStorageInfo storage : QStorageInfo::mountedVolumes())
+	switch (dirType)
 	{
-		storage.refresh();
-
-		if (usableFileSystems.contains(QString(storage.fileSystemType()), Qt::CaseInsensitive) && storage.isValid() && storage.isReady())
-		{			
-			urls << QUrl::fromLocalFile(storage.rootPath());	
-		}
+		case DirType::Project:
+			return config->userProjectsDir();
+		case DirType::Midi:
+			return config->workingDir();
+		case DirType::Preset:
+			return config->userPresetsDir();
+		case DirType::Plugin:
+			return config->userVstDir();
+		case DirType::Sample:
+			return config->userSamplesDir();
+		case DirType::Soundfont:
+			return config->userSf2Dir();
+		case DirType::Song:
+			return config->workingDir();
+		default:
+			return config->workingDir();
 	}
-#endif
-
-	setSidebarUrls(urls);
 }
 
-
-QString FileDialog::getExistingDirectory(QWidget *parent,
```

Comment:
Now I know the reason for the `getExistingDirectory` and `getOpenFileName`. They are overloaded function that [forces](https://doc.qt.io/archives/qt-5.15/qfiledialog.html#getOpenFileName) the use of the Qt file browser to keep a consistent user experience (#4298). There are plans on changing all browsers to OS native (#7856). Maybe we should keep these to maintain status quo until #7856 is merged?